### ISSUE TITLE
InputfieldSelect::isOptionSelected() doesn't make a difference between 0 and blank if Field has multiple values

### ISSUE THREAD

**Issue by kixe** [processwire/processwire-issues#226]
### Short description of the issue
Currently it is not possible to use child classes of **InputfieldSelect** with any fields using **zero** as a selectable value if the child class is an instance of **InputfieldHasArrayValue**.

### Expected behavior
Allow to use values of type `int 0` or `string '0'` for multiple select fields (Chechboxes, InputfieldAsmSelect etc.) delimited from values like '' empty string, `NULL` and `FALSE` which should be handled as NOT SET.

### Actual behavior
Missing consistence (handling of values) between storable values in the database (Select Fieldtypes), the return of `InputfieldSelect::isOptionSelected()`, `InputfieldSelect::renderOptions()` and `InputfieldSelect::isEmpty()`

Whereas `InputfieldSelect::isEmpty()` returns false in case of values `int 0` and `string '0'`, which means the value is allowed, the correct rendering of the field fails.

#### Workaround
I could go around with the following hook
```
$this->addHookBefore("InputfieldSelect::render", function ($e) {
   if ($e->object->isEmpty() && $e->object instanceof InputfieldHasArrayValue == true) $e->object->attr('value', array());
});
```

### Suggestion for a possible fix
In the related function [`InputfieldSelect::isOptionSelected()`]( https://github.com/processwire/processwire/blob/master/wire/modules/Inputfield/InputfieldSelect.module#L221)
convert **int** to **string** for comparison 
```
$valueAttr = array_map('strval',$valueAttr);
if(in_array("$value", $valueAttr, true)) return true;
```

* ProcessWire version: 3.0.57
* PHP version: 7.0
* 3rd party modules related to the issue: FieldtypeSelectExtOption

**Comment by ryancramerdesign** [comment]
What's the best way to duplicate this one? I'm using an InputfieldSelect in a module that has "0" as a  selectable option, but it seems to be working fine here. The code is roughly like this:
~~~~~
$f = $this->wire('modules')->get('InputfieldSelect');
$f->attr('name', 'test');
$f->label = "Test";
$f->addOption('');
$f->addOption(0, 'Zero');
$f->addOption(1, 'One');
$f->addOption(2, 'Two');
$f->attr('value', $this->get('testy'));
~~~~~
The blank (unset) and Zero options are independently selectable and value remains after save. Is there something else I need to add to this to observe the issue?

**Comment by kixe** [comment]
- Everything is working as expected in case of **single values** and if used for Module Configuration.
- It doesn't work in case of **multiple values** (Inputfield instanceof InputfieldHasArrayValue) in combination with **page fields**.

Please try out the following.
Create an integer field 'test' where **blank and zero have different meanings**. Then provide a **MultipleSelect** field in the template to modify the value. I used InputfieldCheckboxes in the example.
```
    $f = $this->wire('modules')->get('InputfieldCheckboxes');
    $f->attr('name', 'test');
    $f->label = "Test";
    $f->addOption('');
    $f->addOption(0, 'Zero');
    $f->addOption(1, 'One');
    $f->addOption(2, 'Two');
    $f->attr('value', $page->get('test'));
    echo $f->render();
```
The Inputfield doesn't make a difference between 0 and blank since `PHP::in_array(0, array('')`) returns true if not strict. `PHP::in_array()` is used by` InputfieldSelect::isOptionSelected()`
The option 'Zero' will be checked in both cases (blank, 0).

This behaviour is not consistent to SingleSelect fields and if used for module settings. I mean it should.

**Comment by Toutouwai** [comment]
Just tested this. I couldn't get a checkboxes option to render with `$f->addOption('')` but could reproduce the same issue when adding an option whose value is just a space character: `$f->addOption(' ', 'Blank')`. If a checkbox option with value 0 is also present then the blank option will be selected/deselected according to the value of the 0 checkbox.

As @kixe suggested the fix for this is simple - just needs the strict parameter set to true [here](https://github.com/processwire/processwire/blob/5b45d17991b9de4755a1ab859bf06958812da0f6/wire/modules/Inputfield/InputfieldSelect.module#L330).

```php
return in_array($value, $this->attr('value'), true);
```

**Comment by ryancramerdesign** [comment]
@Toutouwai @kixe Thanks, I think that makes sense. Though I'm worried it might cause side effects in some cases where the same number might be an int or a string depending on the context. I'm not sure if that actually occurs, but it seems possible given that ints passed through POST come through as strings, and a whole lot of single/multi selection modules extend this base InputfieldSelect. What do you think about a manual strict comparison where we compare everything as strings, like this below? Seems like it would still enable the differentiation between blank and 0, without breaking `"2" == 2` and such.
~~~~~
if($this instanceof InputfieldHasArrayValue) { 
  // multiple selection
  $selected = false;
  foreach($valueAttr as $k => $v) {
    $selected = "$v" === "$value";
    if($selected) break;
  }
  return $selected; 
}
~~~~~

**Comment by Toutouwai** [comment]
Sounds good to me, thanks.

**Comment by kixe** [comment]
Not tested. But as far as I can see, this should work.

**Comment by netcarver** [comment]
@ryancramerdesign Bumping this for your consideration.

**Comment by ryancramerdesign** [comment]
@netcarver It looks like this was already added some time back. Thanks. 

If anyone finds the implemented solution isn't working for them, please let me know, otherwise I think the issue is okay to close.

**Comment by netcarver** [comment]
Right, closing then. If anyone finds this is still an issue, please post a comment.

**Comment by kixe** [comment]
Solved since PW 3.0126
https://github.com/processwire/processwire/commit/2c3e949d6ef836ef616a6636c345d1b740312f2b#diff-79808b1661a74c0668ee6157537ae478

Thanks!

**Comment by kixe** [comment]
Solved since PW 3.0.126
https://github.com/processwire/processwire/commit/2c3e949d6ef836ef616a6636c345d1b740312f2b#diff-79808b1661a74c0668ee6157537ae478
Thanks.
